### PR TITLE
Take incremental nullability snapshot for local functions

### LINKED ISSUE #65061: Assertion in nullable walker in speculation on switch expression
Stack:

I had this code:

```c#
using System;

int? Test1(string str)
{
    return str switch
    {
        "A" => 1,
        "B" => 2,
        "C" => 3,
        _ => default,
    };
}
static int? Test2(string str)
{
    switch (str)
    {
        case "A":
            return 1;
        case "B":
            return 2;
        case "C":
            return 3;
        default:
            return default;
    }
}
```

And i was trying to run `convert switch statement to expression` on the statement in `Test2`.

```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------




   at Microsoft.CodeAnalysis.CSharp.NullableWalker.Variables.CreateNestedMethodScope(MethodSymbol method) in C:\github\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.Variables.cs:line 217

   at Microsoft.CodeAnalysis.CSharp.NullableWalker.GetOrCreateNestedFunctionVariables(Variables container, MethodSymbol lambdaOrLocalFunction) in C:\github\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 2966

   at Microsoft.CodeAnalysis.CSharp.NullableWalker.AnalyzeLocalFunctionOrLambda(IBoundLambdaOrFunction lambdaOrFunction, MethodSymbol lambdaOrFunctionSymbol, LocalState state, MethodSymbol delegateInvokeMethod, Boolean useDelegateInvokeParameterTypes, Boolean useDelegateInvokeReturnType) in C:\github\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:line 3003

   at Microsoft.CodeAnalysis.CSharp.NullableWalker.VisitLocalFunctionStatement(BoundLocalFunctionStatement node) in C:\github\roslyn\src\Compilers\CSharp\Portable\FlowAnalysis\NullableWalker.cs:......

<truncated>
---------------------------
Abort   Retry   Ignore   
---------------------------
```

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes #65061.

Takes incremental snapshot for local functions, so they can be speculated with. Previously because the snapshot was missing, variables for wrong local function were obtained here:

https://github.com/dotnet/roslyn/blob/3b3124564a963dae015846ecb481859da8416138/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L1523

That in turn caused this assert to fail:

https://github.com/dotnet/roslyn/blob/368fb1e1fc745583b57d10aad49cfdf56156d49d/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs#L217-L219

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -63346,6 +63346,51 @@ void f4(object? x4)
...
+                #nullable enable
+                void func() { object? o = null; }
+                void target() { object? o = null; }
+                """;
+            var source2 = """
+                void target() { object? o = ""; }
+                """;
+
+            var comp1 = CreateCompilation(source1).VerifyDiagnostics(
+                // (2,6): warning CS8321: The local function 'func' is declared but never used
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:63362]: >// (2,6): warning CS8321: The local function 'func' is declared but never used [](http://example.com/codeflow?start=16&length=78)

Could we suppress warnings with pragma to reduce the noise? Or they are needed for the repro?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit1)

**Comment by AlekseyTs** [comment]
@333fred Please review

**Review by 333fred — APPROVED** [review]
LGTM (commit 1), modulo Aleksey's `#pragma` comment.

**Pull request merged**